### ISSUE TITLE
Track SmolLM-135M CoreML blockers

### ISSUE THREAD

**Issue by matthargett** [rustnn/rustnn#222]
## Scope

Reproduced with the SmolLM-135M WebNN export against #219 at `13f1ca9` on an M4 Mac.

The graph contains 26 distinct operations. Of those, the CoreML converter already recognizes 25; the only initially unimplemented operation was the internal `shape` interchange op, used 127 times. The recognized set is: `add`, `cast`, `concat`, `cos`, `div`, `equal`, `expand`, `gather`, `greater`, `matmul`, `mul`, `neg`, `pow`, `reduceMean`, `reshape`, `scatterND`, `sigmoid`, `sin`, `slice`, `softmax`, `sqrt`, `transpose`, `triangular`, `unsqueeze`, and `where`.

## Checklist

- [x] Lower internal `shape` nodes to MIL `shape` with CoreML's int32 proxy for WebNN int64 shape tensors (`src/converters/coreml_mlprogram.rs`) — #223.
- [x] Preserve shape-tensor type and rank through `unsqueeze` during WebNN text import (`src/webnn_json.rs`) — #223.
- [x] Cast integer `where` conditions to uniquely named MIL bool values so comparison outputs are not redefined (`src/converters/coreml_mlprogram.rs`) — #223.
- [x] Emit CoreML shape ranges for dynamic graph inputs/outputs and pass the active concrete tensor shape at dispatch.
- [x] Preserve dynamic shapes through `scatterND`, `slice`, and `triangular`, and distinguish scalars from zero-extent tensors.
- [ ] Define the first-token zero-length KV-cache behavior. The proposed WebNN semantics are tracked in [webmachinelearning/webnn#391](https://github.com/webmachinelearning/webnn/issues/391); CoreML flexible-shape ranges still have a minimum extent of 1.
- [x] Bisect the full SmolLM graph's nonzero dynamic-resize failure to maximum extents introduced by synthesized CoreML helper tensors. Gather index normalization expanded active `[1, 1]` indices to `[8, 4096]`; preserving its dynamic MIL dimensions reduces the embedding/projection result from 75,497,472 bytes to the expected 2,304 bytes — [fork PR #9](https://github.com/rebeckerspecialties/rustnn/pull/9).
- [x] Build dynamic `reshape` shapes from runtime MIL `shape` values instead of maximum-size immediates. The first query/key reshapes then return the expected active `[1, 1, 9, 64]` and `[1, 1, 3, 64]` results — [fork PR #10](https://github.com/rebeckerspecialties/rustnn/pull/10).
- [x] Build dynamic `expand`/MIL `tile` repetitions from runtime symbol values. The first rotary expand now returns the expected active `[1, 32, 1]` (128 bytes) instead of `[8, 32, 1]` (1,024 bytes) — [fork PR #10](https://github.com/rebeckerspecialties/rustnn/pull/10).
- [x] Continue the dependency-closed bisect beyond `expand`; the cosine and second multiply branches now return their expected active sizes before the next boundary.
- [ ] Build dynamic `slice` sizes from active runtime dimensions instead of maximum extents. Layer 0 `Slice_3` requests 4096 elements from an active length-1 axis and CoreML rejects it with error `-6`.
- [ ] Continue the dependency-closed bisect through the rest of layer 0 and then the full graph after fixing dynamic `slice`.
- [ ] Complete the remaining dispatch-time shape propagation/specialization needed by [webmachinelearning/webnn#883](https://github.com/webmachinelearning/webnn/issues/883) and the [dynamic-shape proposal](https://github.com/webmachinelearning/webnn/pull/945).
- [ ] Add `shape()`, shape-derived `slice`/`concat`, repeated-dispatch, and zero-extent conformance cases to the [WPT WebNN suite](https://github.com/web-platform-tests/wpt/tree/master/webnn) when the corresponding normative work in #391/#945 lands; keep feature-gated RustNN regressions in the meantime.
- [ ] Isolate the four-token TinyStories-1M numerical failure on Watch SE2 (watchOS 11.6.2, CPU-only). Fixed and dynamic prefill both fail; an Apple-compiled control reproduces the same wrong logit, while one-/two-token cases pass. The four-token cases pass on M4, iPhone XS Max, and Watch Series 10. Compare compute-unit configurations and bisect the first divergent operation; keep this separate from SmolLM's dynamic-slice failure. See [exporter validation](https://github.com/rebeckerspecialties/mlmodelc-export/pull/1).
- [ ] Run one generation step and compare the output token/logits with a reference backend.

## Current diagnostic evidence

On the M4, the current local work produces exact values for fixed ReLU, flexible ReLU at lengths 1/2/3, flexible concat at past lengths 0/1/2, and `shape -> slice -> concat` at lengths 1/2/3. A separately compiled flexible-ReLU `.mlmodelc` also loads and returns the same exact values. Flexible ReLU at length 0 fails prediction with CoreML error `-1`.

A dependency-closed probe now compiles and executes any named SmolLM intermediate with active past length 1. It found three distinct maximum-extent lowerings before the next failing boundary:

1. The embedding gather and query projection declared dynamic shapes but returned their maximum `[8, 4096, 576]` extent because the converter made the gather index-normalization helpers static. Preserving the source index dimensions fixes both active result sizes.
2. The first attention reshape used `[8, 4096, 9, 64]` as an immediate shape. Constructing the shape from runtime dimensions fixes the query and key reshape probes.
3. The first rotary expand used maximum batch 8. Constructing its target shape and MIL `tile` repetitions from live dimensions fixes the expand and advances the probe through the downstream cosine and multiply branches.

The next failing boundary is `_model_layers_0_self_attn_Slice_3_output_0`. Its `slice` lowering requests the maximum size 4096 from an axis whose active length is 1, and CoreML rejects prediction with error `-6`. The complete graph still reports aggregate error `-7`, so the next implementation step is runtime size construction for dynamic `slice`, followed by the same probe through the remainder of layer-0 attention and the complete graph.

Current full-model failures:

```
past length 0: prediction failed with CoreML error -1
past length 1: Error in dynamically resizing for sequence length (-7)
```

**Comment by theHamsta** [comment]
I started to draft the API proposed in https://github.com/webmachinelearning/webnn/pull/945  (very early stage still). I hope that some better shape inference (symbolic and static) will help backends. The changes to the shape inference are still all TBD. I also haven't decided yet whether I will do symbolic shape inference or just the mandated static one for concrete fixed shapes before dispatch.
